core/runtime/jsruntime.rs

```diff
@@ -1304,12 +1304,28 @@ impl JsRuntime {
   pub async fn call_and_await(
     &mut self,
     function: &v8::Global<v8::Function>,
+  ) -> Result<v8::Global<v8::Value>, Error> {
+    self.call_with_args_and_await(function, &[]).await
+  }
+
+  /// Call a function with args. If it returns a promise, run the event loop until that
+  /// promise is settled. If the promise rejects or there is an uncaught error
+  /// in the event loop, return `Err(error)`. Or return `Ok(<await returned>)`.
+  pub async fn call_with_args_and_await(
+    &mut self,
+    function: &v8::Global<v8::Function>,
+    args: &[v8::Global<v8::Value>],
   ) -> Result<v8::Global<v8::Value>, Error> {
     let promise = {
       let scope = &mut self.handle_scope();
+      let mut local_args: SmallVec<[v8::Local<v8::Value>; 32]> =
+        SmallVec::with_capacity(args.len());
+      for v in args {
+        local_args.push(v8::Local::new(scope, v));
+      }
       let cb = function.open(scope);
       let this = v8::undefined(scope).into();
-      let promise = cb.call(scope, this, &[]);
+      let promise = cb.call(scope, this, &local_args);
```

Comment:
I think you can change this to avoid the `SmallVec` when args is empty. It would look something like this:

```
let cb = function.open(scope);
let this = v8::undefined(scope).into();
let promise = if args.is_empty() {
      let mut local_args: SmallVec<[v8::Local<v8::Value>; 32]> =
        SmallVec::with_capacity(args.len());
      for v in args {
        local_args.push(v8::Local::new(scope, v));
      }
      cb.call(scope, this, &local_args)
} else {
  cb.call(scope, this, &[]);
}
```

Reply:
Makes sense! Updated in https://github.com/denoland/deno_core/pull/341/commits/6d48ae327413faf7d3dc9f0ba6c56023c39a3349

Reply:
32 seems excessive here - this is copied from "op code" that expects there's gonna be up to 16 ops returned in one go. In reality I don't expect many situation where embedders would pass more than 8 args to a function.

Reply:
Makes sense. Thanks for the explanation!